crates/bevy_ecs/src/bundle/info.rs

```diff
@@ -543,12 +624,96 @@ impl Bundles {
+        else {
+            // no bundle with `requiree` exists
+            return Ok(false);
+        };
+
+        // `EntityWorldMut::remove_with_requires` generate edges between archetypes where the required
+        // components matter for the the target archetype. These bundles cannot receive new required
+        // components as that would invalidate these edges.
+        // The mechanism is using `Self::contributed_bundle_ids` to track the bundles so we check here if it
+        // has any overlapping `BundleId`s with `bundles_with_requiree`.
+        // TODO: Remove this error and update archetype edges accordingly when required components are added
+        if bundles_with_requiree.iter().any(|bundles_with_requiree| {
+            self.contributed_bundle_ids
+                .values()
+                .any(|bundle: &BundleId| bundles_with_requiree == bundle)
+        }) {
+            return Err(RequiredComponentsError::RemovedFromArchetype(requiree));
+        }
+
+        Ok(true)
+    }
+
+    /// Updates the required components of bundles that contain `requiree`.
+    ///
+    /// # Safety
+    ///
+    /// The caller must have confirmed [`Self::verify_to_refresh_required_components`] returned `Ok(true)`
+    /// for this `requiree` and must pass the same `storages` and `components` to this method that were used
+    /// to create the stored bundles.
+    pub(crate) unsafe fn refresh_required_components(
+        &mut self,
+        storages: &mut Storages,
+        components: &Components,
+        requiree: ComponentId,
+    ) {
+        // take list of bundles to update `Self::bundles_with_requiree` while iterating this
+        let taken_bundles_with_requiree = self
+            .components_in_bundles
+            .get_mut(requiree.index())
+            .filter(|bundles| !bundles.is_empty())
+            .map(core::mem::take)
+            .expect("verify_to_refresh_required_components confirmed existing bundles to refresh");
+
+        for bundle_id in taken_bundles_with_requiree.iter() {
+            let bundle_info = self.bundle_infos.get_mut(bundle_id.index());
+
+            // SAFETY: BundleIds in Self::components_in_bundles are valid ids of Self::bundle_infos
+            let bundle_info = unsafe { bundle_info.debug_checked_unwrap() };
+
+            // SAFETY: Caller ensured `storages` and `components` match those used to create this bundle
+            unsafe {
+                bundle_info.reset_required_components(
+                    storages,
+                    components,
+                    &mut self.components_in_bundles,
+                );
+            }
+        }
+
+        // put the list of bundles with `requiree` back
+        let replaced = self.components_in_bundles.get_mut(requiree.index());
+
+        // SAFETY: this list has to exist at this index to be taken in the first place
+        let replaced = unsafe { replaced.debug_checked_unwrap() };
+
+        *replaced = taken_bundles_with_requiree;
```

Comment:
I would add a note on why this is ok, specifically why this can't overwrite data that was written in the above `reset_required_components`.